spawners and wlu rework for mf

Huckleberry87 opened the pull request:
Extreme Asset Reduction and v1.00.03 update.
- Reduced Pleiku and WLU footprints. Gave WLU a place to call home with some small memorials and some unit flair.
- Added and carved out spawners for WLU and WL locking for vehicles across the board.
- Even Dac base got some 1.3 love, and made AA man a non player.
- Added TP to Satan's Angels hangars, added 633rd to TP list. Reworked TP list to be more intuitive.
- Carrier still on map with reduced assets.
- Many more minor modifications.

Huckleberry87 committed:
spawners and wlu rework for mf

Extreme Asset Reduction and v1.00.03 update.
- Reduced Pleiku and WLU footprints. Gave WLU a place to call home with some small memorials and some unit flair.
- Added and carved out spawners for WLU and WL locking for vehicles across the board.
- Even Dac base got some 1.3 love, and made AA man a non player.
- Added TP to Satan's Angels hangars, added 633rd to TP list. Reworked TP list to be more intuitive.
- Carrier still on map with reduced assets.
- Many more minor modifications.

Huckleberry87 commented:
@wescottj98 this is the same pull request for before, but the merge conflicts have been resolved thanks to @dijksterhuis!

dijksterhuis commented:
reminder: we won't be able to run the KheSahn mission on MF2 until we also edit that mission in editor

wescottj98 reviewed (changes requested):
Please change the following

- string table entries for spawns
- there's a 633rd bird not spawning
- satan's angel bird din't spawn for some reason
- a bit more 'extensive' documentation 
- MACV humming bird spawns in with default skin -- white/grey digital skin not the green one

Huckleberry87 committed:
30MAR PR Fixes

- Fixed string table entries for spawns.
- Fixed 633rd bird not spawning (case sensitive issue in the Respawn Info)
- Fixed Satan's Angel bird not spawn for some reason. The spawner didn't like the hangers. Changed to open top bay like everything else.
- Currently working on a bit more 'extensive' documentation
- This is an issue with how the spawn is randomized via skins (Cerebral/Cypher confirmed): MACV humming bird spawns in with default skin -- white/grey digital skin not the green one.

Huckleberry87 commented:
[MF v.1.00.03 incoming changes.docx](https://github.com/Bro-Nation/Mike-Force/files/14813755/MF.v.1.00.03.incoming.changes.docx)

For publication. Signed off by Polter

dijksterhuis commented:
@Huckleberry87 i knew there was something that was bugging me with this PR........

Where are the new vehicle management scripts....?

Cos we ain't merged those previously and they ain't in your PR ;)

Huckleberry87 commented:
> Where are the new vehicle management scripts....?


@dijksterhuis What do you mean?

dijksterhuis commented:
✅ RESOLVED

---

> > Where are the new vehicle management scripts....?
> 
> @dijksterhuis What do you mean?

In [155](https://github.com/Bro-Nation/Mike-Force/pull/155) there are a bunch of new and old files deleted which were necessary for the vehicle spawning logic.

Where are these files in your pull request?

- https://github.com/Savage-Game-Design/Mike-Force/tree/development/mission/functions/systems/vehicle_asset_manager
- https://github.com/Savage-Game-Design/Mike-Force/tree/development/mission/functions/systems/vehicle_creation_detection

See below from the [files changed tab of 155](https://github.com/Bro-Nation/Mike-Force/pull/155/files)

![image](https://github.com/Bro-Nation/Mike-Force/assets/11841332/d3a7e583-7416-48df-ae5a-63e3fa5d8630)

Huckleberry87 committed:
Client/Server/Global Folders added

Not sure why these 3 subfolders weren't included in my pull request, or even my forked version. Grabbed these from my local working version. I continue to learn. Sorry for the delay.

Huckleberry87 commented:
@dijksterhuis folders added per change request.

dijksterhuis commented:
have pulled locally, will start doing some manual testing over next few days.

Huckleberry87 committed:
Create icons.inc

File was missing

dijksterhuis commented:
RESOLVED

---

@Huckleberry87 I merged https://github.com/Bro-Nation/Mike-Force/pull/271 and forgot it also modified teleporters / the Black Horse 105 white list script.

Make sure you pull `development` then merge into your branch, then fix up those changes. Should be minimal fixes (I think it's basically the same as your own changes).

A merge conflict gets displayed sort of like this in files:

```
some_code = that isnt CONflicting {_x this doesn't need editing};

<<<<<<<< HEAD

THIS SECTION IS THE STUFF THAT YOU HAD

=====================

THIS SECTION IS THE STUFF I JUST MERGED

<<<<<<<<<<<<<<<<<<< development

some_code = that isnt CONflicting {_x this doesn't need editing};

```

You'll need to pick one of the sections between the <<<<< ===== >>>>> and delete those characters and delete the other section. 90% sure they should be the same, or at least very similar.

https://www.atlassian.com/git/tutorials/using-branches/merge-conflicts

Then, to complete the merge, you need to commit and it should carry on merging for you.

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
-	respawnType = "RESPAWN";
-	time = 10;
-};
+#include "..\ui\icons.inc"
```
this file is missing in the branch, needs committing otherwise server will never boot up (@Huckleberry87 has said he's on it)

dijksterhuis replied:
resolved

Huckleberry87 committed:
Merge branch 'development' into camlaonam-vehicle-spawner-wlubases-rework

dijksterhuis commented:
✅ RESOLVED
---------

Need to edit `mission/config/subconfigs/rehandler.hpp` and add this config class, else players cannot select vehicles at a spawn point.

```sqf
class change_spawn_point_vehicle
{
	fnc = "vn_mf_fnc_veh_asset_handle_change_vehicle_request";
};
```

preferably in same manner that SGD did so we don't end up with a future conflict (i.e. second class in the file.)

https://github.com/Savage-Game-Design/Mike-Force/blob/fb22c2fbb22c3692db8c3af81742a5bdcf23f714/mission/config/subconfigs/rehandler.hpp#L5-L8

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+		};
+	};
+
+class acr {
```
style:

```
class acr {
```

to 

```
       class acr {
```

very minor thing.

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+	class vn_b_air_ch34_03_01 {
+		tags[] = {"ch34","armed","transport","helicopter","usmc"};
+	};
+	class C_Heli_Light_01_civil_F {
+		tags[] = {"hummingbird", "transport", "helicopter", "cia"};
+	};
```
wrong class -- needs to be `B_Heli_Light_01_F`

https://community.bistudio.com/wiki/Arma_3:_CfgVehicles_WEST 

https://community.bistudio.com/wiki/File:arma3-b_heli_light_01_f.jpg

edit -- the picture in ^ is wrong. the `B_Heli_Light_01_F` variant does not come with armaments

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_lock_info.hpp:
```diff
@@ -1,26 +1,104 @@
+};
+
+// WL OPFOR
+class dac_cong
```
use this class to lock the MACV birds

dijksterhuis replied:
- 2x air america
- 1x humming bird (MH-9 variant)

dijksterhuis replied:
can also rename the class to `dac_cong_or_macv` to make it clearer to #futureperson what this is (I will forget)

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+				icon = VEHICLE_ICON_CAR;
+				include[] = { { "m151","transport" } };
+				exclude[] = { { "armed","airport","usmc" } };
+			};
```
this submenu is bugged for some reason. the submenu for specific vehicle options does not open. still investigating, but i suspect there are too many options available (maximum possible selections are 8x).

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+				icon = VEHICLE_ICON_ARMOUR;
+				include[] = { { "m113","transport","firesupport","artillery" } };
+				exclude[] = { { "pavn" } };
+			};
```
this submenu is bugged for some reason. the submenu for specific vehicle options does not open. still investigating, but i suspect there are too many options available (maximum possible selections are 8x).

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+				icon = VEHICLE_ICON_TRUCK;
+				include[] = { { "truck","transport","firesupport" } };
+				exclude[] = { { "pavn" } };
+			};
```
this submenu is bugged for some reason. the submenu for specific vehicle options does not open. still investigating, but i suspect there are too many options available (maximum possible selections are 8x).

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+				name = "UH-1";
+				icon = VEHICLE_ICON_HELO;
+				include[] = { { "uh1","arvn" } };
+			};
```
this submenu is bugged for some reason. the submenu for specific vehicle options does not open. still investigating, but i suspect there are too many options available (maximum possible selections are 8x).

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+				icon = VEHICLE_ICON_CAR;
+				include[] = { { "m151a1", "armed" } };
+				exclude[] = { { "transport","usmc" } };
+			};
```
this submenu is bugged for some reason. the submenu for specific vehicle options does not open. still investigating, but i suspect there are too many options available (maximum possible selections are 8x).

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_lock_info.hpp:
```diff
@@ -46,221 +124,249 @@ class vn_b_armor_m113_acav_04 : acav {};
 //Howitzer
-class vn_b_army_static_m101_02 : acav {};
-class vn_b_sf_static_m101_02 : acav {};
+class vn_b_army_static_m101_02 : QuarterHorse {};
+class vn_b_sf_static_m101_02 : QuarterHorse {};
```
whitelist is allowing me in as a green hornet for some reason. tested at bersctedhdsjfhsdfgarden

Huckleberry87 replied:
Possible issue with statics as opposed to other vehicles.

dijksterhuis replied:
✅ RESOLVED

dijksterhuis commented on mission/config/subconfigs/vehicle_lock_info.hpp:
```diff
@@ -46,221 +124,249 @@ class vn_b_armor_m113_acav_04 : acav {};
+//Type 63 tank
+class vn_i_armor_type63_01 : arvn {};
+
+// M48 Patton
+class vn_b_armor_m48_01_01 : armor {};
+
+// M67 flame tank
+class vn_b_armor_m67_01_01 : armor {};
+
```
no locking the USMC variants to USMC? `vn_b_armor_m48_01_02` and `vn_b_armor_m67_01_02`

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_lock_info.hpp:
```diff
@@ -46,221 +124,249 @@ class vn_b_armor_m113_acav_04 : acav {};
+class vn_b_armor_m125_01 : armor {};
+class vn_b_armor_m132_01 : armor {};
+
+//M41 Walker bulldog
+class vn_b_armor_m41_01_01 : armor {};
```
no locking the USMC variant to USMC? `vn_b_armor_m41_01_02`

dijksterhuis replied:
resolved in line below

dijksterhuis reviewed (changes requested):
first pass.... will try to add some screenshots in

## scripts / not editor

### urgent: quarter horse 105 + dac cong statics whitelist

- i could get in the 105s as a green hornet ✅ DONE
- i could get in the dac cong arty spawns at dac HQ as a green hornet  ✅ DONE

@Huckleberry87 and I think this is because whitelisting is only functional on actual vehicles, and not for statics.

may need to reintegrate the old QuaterHorse whitelist locking

### whitelist for usmc tanks  ✅ DONE

^

### humming bird macv + whitelist macv (reuse dac cong whitelsit)  ✅ DONE

^

### ui menu glitches  ✅ DONE

@dijksterhuis investigating

you saw these earlier today when i was testing. gonna see if i can work out what is going on.

#### no vehicles spawning  ✅ DONE

- Root cause identified ✅ ... awaiting fix.

Some vehicle class names in the `vehicle_respawn_info.hpp` config do not exist in `CfgVehicles` (they aren't assets in the game known by that name anymore). 

Those classnames were likely in the previous version of the file, leftover from pre-1.3 DLC days where there was no ROK faction etc. once SGD added the ROK faction, quite a lot of vehicles would have gone from `old_class_name` to `old_class_name_rok_01_02`.

To fix -- we need to double check vehicle class names against https://wiki.sogpf.com/index.php/CfgVehicles_WEST

(The wiki is the authorities list of all classnames in the DLC. Bookmark it! )

I can do some funky scripting magic to check for anything outstanding as there are A LOT of classnames we need to verify.

Some spawn points I spotted as being affected.

- small boats udt 1 (closest to the shoreline) -- nothing spawns in
- 7thcav mule/jeep does not spawn at all
- SA jets not spawning on carrier
- SA jet not spawning at pleiku
- SA peliku jet 2 --> trying to  select it casues the bad class error and it does not spawn


### incorrect vehicle names when vehicle is IDLE on map  ✅ DONE

- Root cause identified ✅ ... awaiting fix.

PBRs and other small boats showing as Rearm/Repair, P1, P2, P3 on the map when left idle ... ~~i think this must be a missing entry in the stringtable xml file. not sure.~~

is related to a conflict with existing marker names in the `mission.sqm`. IIRC Arma defaults to naming markers placed in the `mission.sqm` as `marker_number+1`. 

This issue is something upstream (SGD) didn't identify as a potential issue (probably they didn't come across it), but we're hitting due to scale of our mission file.

### non functional vehicle selection submenus  ✅ DONE

- Root cause identified ✅ ... awaiting fix.

the UI wheel menu can only support a maximum of 10 entries. 2 of those are always the earplugs and scout action -- regardless of whether you can perform the scout action or not. regardless of whether they are actually shown in the sub menu or not.

^ that means it is only possible to have 8x entries in the wheel menu.

I counted 14x `m151a` vehicles in the vehicle tagging config.

I think some of the spawns have too many vehicles for certain vehicle groups. and that may be causing the vehicle submenus to not open, as noted on the specific category classes in review.

----

## potential editor / non scripting things

### vehicle spawning issues

#### spawn position placement issue?  ✅ DONE

- ARVN heavy vehicle 2 -- does not switch because of vehicle placement? (the eastern most placement)
- spike team / green hornet overflow 1 (western most out of the 2 at the end) -- does not switch because of vehicle placement?

### WLU bases specific
- ANZAC -- a lamp next to dirt rangers -- difficult to see vehicles at night!  ✅ DONE
- ahem ... ANZAC only get a cayuse? no slick / standard helo spawn? i'm gonna cause a riot.  ✅ DONE
- I reckon tigers will want to move back to plei mei SF camp. it's possible to move them back up there and place the helo spawn on the pad outside the base. 

### misc

- some of the signs could do with being lowered! had to snap my neck all the way back to select some of the vehicles.  ✅ DONE

dijksterhuis commented on mission/functions/systems/vehicle_asset_manager/server/fn_veh_asset_marker_create.sqf:
this function looks up the vehicle's `displayName` from `CfgVehicles` to display the vehicle's name on map.

```
(configFile >> "CfgVehicles" >> typeOf _vehicle >> "displayName")
```

but for small boats, and only when the boats are `IDLE`, the function finds names from places on the map for some reason... is weird. not sure how it is doing that.

### boats are idle -- shows strings from other map marks

![20240406151730_1](https://github.com/Bro-Nation/Mike-Force/assets/11841332/23362ff7-0d4c-4f07-840c-89d2da599104)

### player in boat -- shows the correct string

![20240406172013_1](https://github.com/Bro-Nation/Mike-Force/assets/11841332/6bffea7f-0a7a-4fd5-a8a7-9cffa247a8d8)

dijksterhuis replied:
issue identified. see below.

dijksterhuis commented on mission/functions/systems/vehicle_asset_manager/server/fn_veh_asset_marker_create.sqf:
```diff
@@ -0,0 +1,59 @@
+	}
+];
+
+private _markerName = format ["marker_%1", _spawnPoint get "id"];
+[_markerName, getPos _vehicle, _textGenerator] call para_g_fnc_create_localized_marker;
```
I think there could be a conflict in marker names that needs to be resolved. i'm 97% sure when placing down a map marker in eden editor, Arma will name it using this same convention `marker_1` etc.

I think we may want to change this 
```
private _markerName = format ["marker_%1", _spawnPoint get "id"];
```

to something like
```
private _markerName = format ["vehicle_live_marker_%1", _spawnPoint get "id"];
```

but we'll need to check for any other references to the `marker_%1` string in the vehicle system and also change them to `vehicle_live_marker_%1`

dijksterhuis replied:
confirmed is a a conflict with `mission.sqm` marker names

```
C:\Users\dijksterhuis\Documents\MIKE-FORCE\Mike-Force\maps\cam_lao_nam\mission.sqm:
 13289  			dataType="Marker";
 13290  			position[]={17184.65,14.97998,7022.7148};
 13291: 			name="marker_100";
 13292  			text="SP1";
 13293  			type="mil_warning";
 .....
 13913  			dataType="Marker";
 13914  			position[]={20067.029,1.4619141,9779.8848};
 13915: 			name="marker_101";
 13916  			text="P2";
 13917  			type="mil_pickup";
 .....
 13925  			dataType="Marker";
 13926  			position[]={20083.482,1.1785889,9773.7529};
 13927: 			name="marker_103";
 13928  			text="P3";
 13929  			type="mil_pickup";
 .....
 15777  			dataType="Marker";
 15778  			position[]={17423.734,30.685818,5639.2544};
 15779: 			name="marker_124";
 15780  			text="Rearm/Repair";
 15781  			type="b_maint";
 .....
```

dijksterhuis replied:
for reference


### boats are idle -- shows strings from other map marks

![20240406151730_1](https://github.com/Bro-Nation/Mike-Force/assets/11841332/23362ff7-0d4c-4f07-840c-89d2da599104)

### player in boat -- shows the correct string

![20240406172013_1](https://github.com/Bro-Nation/Mike-Force/assets/11841332/6bffea7f-0a7a-4fd5-a8a7-9cffa247a8d8)

dijksterhuis replied:
✅ RESOLVED

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+	class vn_b_air_f100d_at {
+		tags[] = {"f100","armed","firesupport","jet"};
+	};
+	class vn_b_air_f4b_at {
+		tags[] = {"f4","armed","firesupport","jet","naval"};
+	};
```
this class does not exist -- this is what is causing vehicles to not spawn.

use this to double check vehicle class names -- https://wiki.sogpf.com/index.php/CfgVehicles_WEST

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+	};
+	class vn_b_wheeled_m151_mg_01 {
+		tags[] = {"m151a1","armed","firesupport","car"};
+	};
```
This class does not exist. This is causing spawn failures for certain spawn points. Check vehicle class names here -- https://wiki.sogpf.com/index.php/CfgVehicles_WEST

Huckleberry87 replied:
resolved

dijksterhuis commented on mission/config/subconfigs/vehicle_respawn_info.hpp:
```diff
@@ -1,491 +1,1606 @@
+	class vn_o_armor_t54b_01 {
+		tags[] = {"t54","pavn","armor"};
+	};
+	class vn_o_armor_ot54b_01 {
+		tags[] = {"t54","pavn","armor"};
+	};
```
Vehicle class does not exist. ...... see other comments about the issue.

Huckleberry87 replied:
resolved

Huckleberry87 committed:
PR CR 1

Addressed fixes needed by DJ as of 06APR.

m151 parsed down, f4b changed to add Navy, ot54b - removed the b, usmc tanks added to lock list, arvn uh1 added, acr m113s parsed down.

vehicle marker name line

Huckleberry87 committed:
PR CR 2

Changed humming bird lock info

Huckleberry87 committed:
PR CR 3

Removed deprecated vehicles from lock and spawner.

Altered functions (vehicle info lines) based on offsite change request.

Huckleberry87 committed:
PR CR 4

Removed Cayuse from MACV Spawn. Changed Hummingbird class name. Parsed out the jeeps and 113s a bit better to reduce spawn selection. Should address the issue with "too many choices" for the wheel selection menu.

Huckleberry87 committed:
PR CR 5

-Move some vehicle spawns around to allow them to spawn in.
-Changed aircraft spawner at ANZAC base to allow for uh1/ch47.
-Added lights to ANZAC base.
-Added lights to FSB Berch.
-Lowered all the sign posts to be easier to access.
-Organized all the stuff into folders for ease of use for editor.
-Fixed m151 spawner for light transport class. (m151, should be m151a1).

Played on my local side and all vehicles are selectable and spawnable.

dijksterhuis committed:
VehiclesRework: also lock statics to defined teams.

dijksterhuis committed:
chore: fixup function name for clarity.

dijksterhuis committed:
VehiclesRework: Use category icon as surrogate for vehicle icon

Vehicles don't have icons data provided in the configs.

It seems like we were supposed to be able to editing vehicle specific respawn configs to have an icon, see the below

```
	class vn_b_air_ch47_01_01 {
		tags[] = {"ch47","armed","transport","helicopter","army"};
        icon = VEHICLE_ICON_CAR;
	};
```

However

- this config ^ doesn't show the icon (it doesn't work)
- it would be tedious when we've already done the same for categories
- vehicle icons will likely just be the same as the category anyway

Instead, we can keep life simple, and just grab the top level category icon as a surrogate icon for each vehicle.

Huckleberry87 committed:
Merge pull request #1 from dijksterhuis/camlaonam-vehicle-spawner-wlubases-rework-static-weps-authz

VehiclesRework: also lock statics to defined teams.

Huckleberry87 committed:
Merge pull request #2 from dijksterhuis/fixup-scroll-wheel-dispaly-vehicle-icons-from-category

VehicleRework: Fixup scroll wheel display vehicle icons from category

dijksterhuis committed:
Add/edit stringtable map marks to be more clear whilst more generic.

dijksterhuis committed:
Mission edits, see commit body for full list of changes

- Move around map marks, attmpt to minimise collision when zoomed out
- Add some additional map marks like Motor Pool and other Supply Drops
- Raise ANZAC helo spawn by a boot to avoid borking the spawn
- Raise Motor pool heavy vehicles to avoid borking during spawn
- Move small boat signs to be next to the boat rather than behind them for better UX

dijksterhuis committed:
chore: tidy up formatting.

dijksterhuis committed:
finallky increment version number :tada:

dijksterhuis committed:
Last mission edit fixup: libation and arty crates at blackhorse.

- MACV + ANZAC + Muskets 'libations' crate were falling through the ground and being destroyed on mission start
- Moved Mortar crates at Blackhorse to inside nearby billets to stop them being destroyed on mission start up

dijksterhuis commented:
@wescottj98 once @Huckleberry87 merges https://github.com/Huckleberry87/Mike-Force/pull/3 into his PR then I'll be happy. ✅

Huckleberry87 committed:
Merge pull request #3 from dijksterhuis/final-mission-edits-and-fixups

Final mission edits and fixups

dijksterhuis commented:
LGTM 🎉

Huckleberry87 committed:
Update vehicle_lock_info.hpp

Removed lock info for a vehicle not on the spawn list. Removed a redundant category comment
Added UH-1B for ARVN that is on spawn list.
Added UH-1D for ARVN that is on spawn list.
Reorganized specific list order because numbers must be sequential (OCD)

cerebral-314 commented:
Gonna push this along... changes seem to be sound.
